plan/attributes.go

```diff
@@ -1,27 +1,81 @@
+//
+// For example, if a node requires its input tables to be sorted, it
+// will have the CollationAttr among its required attributes.
+// Likewise, if a node produces sorted tables, its output attributes
+// will have CollationAttr.
+//
+// Operations can require or provide attributes by implementing interfaces
+// in the corresponding PhysicalProcedureSpec:
+// - OutputAttributer is to be implemented if the procedure will provide an attribiute. E.g.,
+//   the SortProcedureSpec will provide CollationAttr for the columns on which the data is
+//   to be sorted.
+// - PassThroughAttributeris to be implemented if a procedure will not perturb a given attribute.
+//   E.g., if data with CollationAttr flows into a filter, it will still be sorted. Therefore,
+//   FilterProcedureSpec can implement PassThroughAttributer for collation.
+// - RequiredAttributer is to be implemented by procedures that require particular attreibutes.
+//   There is one set of required physical attributes for each input, since they may be different.
+//   E.g., SortMergeJoinProcedureSpec requires that the left and right inputs be sorted on the
+//   columns being joined on (and they may in fact have different names on either side).
+//
+// It's the obligation of planner rules to ensure that required attributes are satisified by
+// a procedure's inputs. If a node has required attribute that are not satisfied, it will be
+// caught by ValidatePhysicalPlan(), and an internal error will be returned.
+
+// PhysicalAttr represents an attribute (collation, parallel execution)
+// of a plan node.
+type PhysicalAttr interface {
+	Key() string
+	SuccessorsMustRequire() bool
+	SatisfiedBy(attr PhysicalAttr) bool
+}
+
+// PhysicalAttributes encapsulates any physical attributes of the result produced
+// by a physical plan node, such as collation, etc.
+type PhysicalAttributes map[string]PhysicalAttr
+
+// OutputAttributer is an interface to be implemented by PhysicalProcedureSpec implementations
+// that produce output that has particular attributes.
+type OutputAttributer interface {
+	OutputAttributes() PhysicalAttributes
 }
 
-// If a node produces parallel data, then all successors must require parallel
-// data, otherwise there will be a plan error.
-func (ParallelRunAttribute) SuccessorsMustRequire() bool {
-	return true
+// PassThroughAttributer is an interface to be implemented by PhysicalProcedureSpec implementations
+// that allow attributes to propagate from input to output.
+type PassThroughAttributer interface {
+	PassThroughAttribute(attrKey string) bool
 }
 
-type ParallelMergeAttribute struct {
-	Factor int
+// RequiredAttributer is an interface to be implemented by PhysicalProcedureSpec implementations
+// that require physical attributes to be provided by inputs. The return value here is a slice,
+// since each input may be required to have a different set of attributes.
+type RequiredAttributer interface {
+	RequiredAttributes() []PhysicalAttributes
 }
```

Comment:
Will every procedure spec have to implement these `*Attributer` interfaces? Or will it be assumed that if a spec doesn't implement `RequiredAttributer` (as an example) that it has no required attributes?

Reply:
Yes, it's the latter case. If there are no required attributes, then no changes need to be made. Likewise for the other interfaces, operations won't produce or pass on any attributes of the interfaces are unimplemented.

Reply:
So in the case of collation, any transformation that doesn't disrupt sorting will will have to implement `PassthroughAttributer` correct?

Reply:
If we want to be able to remove sort nodes that come after it, then yes. It won't be a wrong answer if we forget to implement it, but we might get a slower query.

I think we'll see a limited number of nodes between the sort and a source. It might work something like this for example:
```js
ReadFilter // implements OuptutAttributer with CollationAttr
  |> filter(...) // filter that could not be pushed down; implements PassThroughAttributer for collation
// |> group() // if you have this here it would disrupt collation; no need to implement anything new
  |> sort() // could be removed by a planner rule

```
Regardless of the mechanism, we'll need to look at each transformation and decide if it maintains collation and implement `PassThroughAttributer` for it if so. We definitely don't want to make it a default of any kind.

I can think of other ways to do it, like keeping a list of all the transformations that maintain collation, but this is the most generic way I could think of that could be extended to other kinds of attributes.

Reply:
This is not a strong opinion, and I don't immediately know what it would take to implement this, but I think it might be simpler if we instead had nodes report that they disrupt sorting, rather than reporting that they preserve it. So if a node doesn't explicitly report that it disrupts sorting, we assume sorting remains unchanged. Seems like it would accomplish the same thing and require less boilerplate code.

Again, this is opinion is loosely held, and I'm interested in hearing any pushback to it. Ultimately, this patch seems like a good solution, and I'd be content if this ends up getting merged.

Reply:
> So if a node doesn't explicitly report that it disrupts sorting, we assume sorting remains unchanged.

I have concerns with this idea. What I really want to avoid is a case where an operation needs sorted input but due to programmer error it receives unsorted data and produces a wrong answer. This would happen with merge join for example.

I would much rather we see queries be slow because we are sorting redundantly than get a wrong answer the the customer might not realize.

It could result in some boilerplate, but that could be minimized with a mixin:
```Go
type PassThroughCollation struct{}
func (ptc PassThroughCollation) PassThroughAttribute(attrKey string) bool {
  return attrKey == plan.CollationKey
}

// filter.go

type FilterProcedureSpec struct {
  PassThroughCollation
  // other fields...
}
```

With this approach, we would need to just add a single line to the procedures that pass on collation to successors.

Reply:
>What I really want to avoid is a case where an operation needs sorted input but due to programmer error it receives unsorted data and produces a wrong answer ... I would much rather we see queries be slow because we are sorting redundantly than get a wrong answer the the customer might not realize.

That is a fair point that had not occurred to me. Consider me convinced.

In that case, this patch looks good. Assuming we can resolve the problems with the parallel attribute, I'd be happy to see something like this get merged.

Reply:
Thanks, I'll file an issue with specifics on how to make this production-ready and put it in the epic.